lib/Cabal2Nix.hs

```diff
@@ -272,6 +261,35 @@ toNixPackageDescription isLocal detailLevel pd = mkNonRecSet $
         , "extraDocFiles" $= toNix (extraDocFiles pd)
         ]
 
+srcToNix :: PackageIdentifier -> Src -> NExpr
+srcToNix _ (Path p) = mkRecSet [ "src" $= applyMkDefault (mkRelPath p) ]
+srcToNix pi' (PrivateHackage url)
+  = mkNonRecSet $
+    [ "src" $= applyMkDefault (mkSym pkgs @. "fetchurl" @@ mkNonRecSet
+      [ "url" $= mkStr (fromString $ mkPrivateHackageUrl url pi')
+      , "sha256" $= (mkSym "config" @. "sha256")
+      ])
+    ]
+srcToNix _ (Git url rev mbSha256 mbPath)
+  = mkNonRecSet $
+    [ "src" $= applyMkDefault (mkSym pkgs @. "fetchgit" @@ mkNonRecSet
+      [ "url"    $= mkStr (fromString url)
+      , "rev"    $= mkStr (fromString rev)
+      , "sha256" $= case mbSha256 of
+                      Just sha256 -> mkStr (fromString sha256)
+                      Nothing     -> mkNull
+      ])
+    ] <>
+    [ "postUnpack"
+      $= mkStr (fromString $ "sourceRoot+=/" <> root <> "; echo source root reset to $sourceRoot")
+    | Just root <- [mbPath] ]
```

Comment:
Why not use the `instance ToNixExpr Src where`?

Reply:
Because this function has to use `PackageIdentifier` to construct Hackage URL, and I found no easy way to include that into `Src` type itself. 

`Src` is constructed in `hackage2nix/Main.hs` and cabal file is not parsed yet at that point.

Do you have a better idea how to pass `PackageIdentifier` here?